### PR TITLE
[UnitPrice] Manage comma as decimal separator in unit value field

### LINKED ISSUE #7710: Commas parsed incorrectly, leading to incorrect "Display as" and "Unit Price" values
## Description

Follow-up from issue #7616.

When creating anew product, the "Value" field does not interpret commas `,` correctly. Setting `Value` = `1,5` for `Unit Size *` = `Weight (kg)` is interpreted as `1 kg ,5`; This also leads to incorrect calculation of `Unit Price`.

## Expected Behavior

Commas are expected to be parsed as dots.

## Actual Behaviour

Commas truncate the input on the `Value` field, leading to incorrect `Display as` info throughout the app, as well as incorrect calculation of `Unit Price`. 

## Steps to Reproduce

As admin
1. Create a new product
2. Select `Unit Size *` = `Weight (kg)` for example
3. Type `1,5` on the `Value` field
4. Notice that this interpreted as `1 kg ,5`, as seen on the `Display as` field
5. Type `1` on `Price`
6. Notice `Unit Price` is `1` and not `1,5` as expected

## Animated Gif/Screenshot

image

image

image

When editing the variant;
image

Shopfront
image

## Workaround

- Use dots instead to correctly output the "Unit value" field
- Override the "Display as" field

## Severity

bug-s3: a feature is broken but there is a workaround

## Your Environment

* Version used: v3.7.2
* Browser name and version: Firefox 88
* Operating System and version (desktop or mobile): Desktop Ubuntu 20.04

### PR DESCRIPTION
**Pull request by jibees**
#### What? Why?

Closes #7710

This PR is here to be more and more tolerant when parsing a value inside the unit value field.
As configuration for currency settings has no effect when displaying currency in the frontend (because we're using `i18N` lib which doesn't take into account this configuration), we should handle comma as decimal separator even if the browser language is a one that handle final point as decimal separator.
Capture d’écran 2021-06-08 à 09 53 48

##### As admin, creating a new product
1. Go to `/admin/products/new`
2. Select `Unit Size *` = `Weight (kg)` for example
3. Type `1,5` on the `Value` field
4. Notice that this interpreted as `1,5kg`, as seen on the `Display as` field
5. Type `1,5` on `Price`
6. Notice `Unit Price` is `1.00`
Capture d’écran 2021-06-08 à 10 44 50

##### As admin, editing or creating a product/variant
1. Go to `/admin/products/[PRODUCT_NAME]/variants/[VARIANT_ID]/edit` (or creating new variant url: `admin/products/[PRODUCT_NAME]/variants/new`
2. Type `1,5` on the `WEIGHT (G)` field (for example)
3. Type `1,5` on the `PRICE` field
4. Check that the `Unit Price` field is well filled and its value is mathematically logical
Capture d’écran 2021-06-08 à 10 51 35

#### Release notes
Better management of decimal price in the backoffice around unit prices.
Changelog Category: User facing changes

### CODE REVIEW

**Comment by jaycmb** [comment]
I´m wondering if the solution is to be 

> more tolerant when parsing a value inside the unit value field

In (the rare case) that users have prices that need a thousands separator, it should make a difference if a user enters
1,000.00 or 1.000,00.

This bug seems to be more related to the issue with the [localization](https://github.com/openfoodfoundation/openfoodnetwork/issues/7719) vs. currency settings per instance?

File: app/assets/javascripts/admin/utils/filters/unlocalize_currency.js.coffee [hunk 1]
```diff
@@ -1,15 +1,26 @@
...
-      
+
     if (decimal_separator == ",")
       price = price.replace(",", ".")
 
-    return parseFloat(price)
+    price = parseFloat(price)
+
+    if (isNaN(price))
+      return null
```
Comment (luisramos0) [inline app/assets/javascripts/admin/utils/filters/unlocalize_currency.js.coffee:24]: return null if isNaN(price)

File: app/assets/javascripts/admin/utils/filters/unlocalize_currency.js.coffee [hunk 2]
```diff
@@ -1,15 +1,26 @@
...
     thousands_separator = I18n.toCurrency(1000, {precision: 1, unit: ''}).substring(1,2)
+    
+    # Replace comma used as a decimal separator and remplace by "."
+    if (price.match(/^[0-9]*(,{1})[0-9]{1,2}$/g))
+      price = price.replace(",", ".")
 
-    if (price.length > 4)
-      # remove configured thousands separator if price is greater than 999
+    # Remove configured thousands separator if it is actually a thousands separator
+    if (new RegExp("^([0-9]*(" + thousands_separator + "{1})[0-9]{3}[0-9\.,]*)*$", "g").test(price))
```
Comment (luisramos0) [inline app/assets/javascripts/admin/utils/filters/unlocalize_currency.js.coffee:15]: can you move this regexp test to a method and give it a good name so everyone understands what is the test?

**Review by luisramos0 — APPROVED** [review]
:+1:

File: app/assets/javascripts/admin/utils/filters/unlocalize_currency.js.coffee (hunk 2 above)
Comment (jibees) [reply app/assets/javascripts/admin/utils/filters/unlocalize_currency.js.coffee]: Thats ok! See db8f8a267

**Comment by filipefurtad0** [comment]
Hey @jibees ,

this was tested on staging-AU, using the EN locale and these settings:

image

We can see that after this PR the input accepts commas as decimal separators - despite being set otherwise:

Peek 2021-06-11 10-05

The "Display as" is displayed in several places throughout the app, such as when setting an **order cycle**:

image

**Shopfront**
image

Editing a variant
image

We can see that the newly set variant is displayed correctly, but also that the variants which were displaying the bug were "corrected" after deploying this PR! :tada: Compare the `display_as_product ` - before this PR (see the [pics on the issue](https://github.com/openfoodfoundation/openfoodnetwork/issues/7710#issue-906424154)) with the pics above.

The question is however what happens when `we mix both` - @jaycmb makes a good point:
> it should make a difference if a user enters 1,000.00 or 1.000,00.

Or, before that: `when does a comma stops being a decimal point to become a thousands separator`? :eyes: 

This depends on what comes _right_ to the comma - which is rather unusual. For example:

- `1,5` is parsed as 1.5 (one and a half)
- `1,50` is also parsed as 1.5 (one and a half)
- `1,500` is also parsed as 1500 (thousand five hundred)

Peek 2021-06-11 10-44

When it comes to mixing both:

- `1,000.00` is parsed as 1000
- `1.000,00` is parsed as 1

However, and as already showed:
- `1000,00` is parsed as 1000 -> why? only two digits to the right
- `1,00000` is parsed as 100000 -> why? over two digits to the right

It could be interesting to double check this on a server with the "opposite" settings as staging-AU.

**Summary**
So, it turns out that for the settings on this server:
- a comma stops being a decimal separator after three digits to its right (this is new math!)

I think this is a good compromise... Ideally, users will know and use the settings in their country/region. Although this will surely not always be the case, it is probably out of scope for OFN: this PR prevents unit prices from breaking, and follows a tangible logic - even if not strictly following arithmetic conventions .

So I'd say this good to go. But I'd like to know what you @jibees and @jaycmb thinks about this.

Again, awesome effort JB, also with all the specs :clap: :clap: :clap:

**Comment by jaycmb** [comment]
In general I was a bit concerned about "equalizing"` ,` and `. `, because 

a) this leads to some confusion and potentially problems at other parts of the product, when these inputs are processed that would be sensitive to `,` or `.` being treated distinct
-> to be checked by @filipefurtad0 (or @jibees to confirm if the change of `price `or `value` input here is only UI related)

b) Also, needs to be checked how this applies to instances with `,` being decimal separator, because

>  1000,00 is parsed as 1000 -> why? only two digits to the right
>     1,00000 is parsed as 100000 -> why? over two digits to the right

would be  **not correct** for EU instances where  1,00000 should be parsed as 1

In general, I think a solution that prevents the user from entering` value` or `price` in alignment with the localization settings would be more favorable instead of making the input fields more tolerant, like displaying an error message _"Please enter this value according to the local number settings with `.`as thousands and `,`as decimal separator._ before saving the product.

After discussion with Filipe, we clarified the following questions

1. No matter if user chooses 1,5 or 1.5
- Display As and Unit Price will be 1.5?
- What happens if user enters , for Value but . for Price (or vice versa?)

-> The decimal separator that has been chosen on an instance level will be the used for _display as_, _unit price_ and shopfront

2. Will we encounter mixed displays across different parts of the product?
-> Possibly, see the [localisation issue #7719](https://github.com/openfoodfoundation/openfoodnetwork/issues/7719) for more info

**Overall seems like this PR**
- prevents unit prices from breaking
- The instance settings will overwrite whatever the user enters and display it according to the instance´s number conventions, so we won´t encounter shopper facing discrepancies
- We assume it´s a rather rare case that prices and quantities above 1000 will be entered, which are the critical cases

So if a) and b) are checked and confirmed, I d say good to go to not put more time and resources in this one for now.
But certainly the [localisation issue](https://github.com/openfoodfoundation/openfoodnetwork/issues/7719) needs attention at some point.

**Comment by jibees** [comment]
@jaycmb 
> or @jibees to confirm if the change of price or value input here is only UI related

My modifications are UI related: the unit price field _only_ revealed these inconsistencies. So i guess this behavior should be limited to the unit price scope.


> In general, I think a solution that prevents the user from entering value or price in alignment with the localization settings would be more favorable instead of making the input fields more tolerant, like displaying an error message "Please enter this value according to the local number settings with .as thousands and ,as decimal separator. before saving the product.

I think it's a great setback, thanks for that @jaycmb. I'm afraid that It could lead to some other bugs, related to the inconsistances between instance configuration and browser configuration, but maybe worth the effort at some point.


> > 1,00000 is parsed as 100000 -> why? over two digits to the right

>would be not correct for EU instances where 1,00000 should be parsed as 1

Not sure to understand: should I care for this one, and modify anything to take into account this case? I'd happy (_not really, but yeah ;)_) to manage this one, let me know.

**Comment by jaycmb** [comment]
> 

        1,00000 is parsed as 100000 -> why? over two digits to the right

    would be not correct for EU instances where 1,00000 should be parsed as 1

> Not sure to understand: should I care for this one, and modify anything to take into account this case? I'd happy (not really, but yeah ;)) to manage this one, let me know.

This needs do be tested first how this is currently handled for EU instances with localisation settings set to `,` being decimal separator. Maybe does not need to be addressed because it´s automatically handled.

With the new bug that came up (discussed in Slack), 
> that if users input any letters into the unit field then it messes up the weight calculation.

we might need to consider the solution that prevents users from entering incorrect inputs (prevent product from being saved and display error message), but understood that 

>  inconsistences between instance configuration and browser configuration
 need to be investgated

**Comment by jaycmb** [comment]
Checked for France:
Entering 1.000 (thousand, according to number localisation) grams is parsed as 1g (leading to a Unit price of 1 500,00 € where it should be 1,50€
image

**Comment by jibees** [comment]
> Entering 1.000 (thousand, according to number localisation)

On your screenshot, this seems to be `1,000` right?

**Comment by filipefurtad0** [comment]
Some issues I think we should consider, to clarify this a bit, as we have an overlap of different issues - these are not introduced by this PR:

i) there is a discrepancy throughout the app, on how Currency/Localization logic is applied. Shopfront displays commas as decimal separator and backoffice does not (see this comment). In some cases, this discrepancy is even displayed on the same screen, or on closely related screens.

ii) choosing a different language on your browser session can change Currency/Localization logic, in some places in the app but not all.

iii) parsing on the relevant fields for this issue are unaffected by i) and ii), so the tests performed before on staging-AU are valid and confirmed now for staging-ES

iv) as far as I could gather, the relevant fields are:
a) "Value" `product_unit_value` - when creating a new product on `/admin/products/new`
b) "Weight" or "Volume" `unit_value_human` - when editing a variant on `/admin/products/<prod>/variants/<nr>/edit`
c) "Unit" `variant_unit_value_with_description` - when editing a variant on `/admin/products`

This PR changes the behavior on fields b) and c) only.

Field a) does not break unit price nor triggers the [related NaN error](https://openfoodnetwork.slack.com/archives/CEF14NU3V/p1623840052097300), because the input is truncated after an invalid character is typed.

This PR prevents the b) and c) from breaking due bad parsing of commas. Also, the issue found earlier today which seems to relate to typing letters (like kg) is fixed as well by this PR :+1

Before:
Peek 2021-06-16 12-35

After:
Peek 2021-06-16 13-04

**Comment by RachL** [comment]
> Checked for France:
Entering 1.000 (thousand, according to number localisation) grams is parsed as 1g (leading to a Unit price of 1 500,00 € where it should be 1,50€


Just FYI. That scenario for FR is correct as the unit is grams. I wouldn't change this behavior.

**Comment by filipefurtad0** [comment]
> In general I was a bit concerned about "equalizing" , and . , because

I share @jaycmb concern though, especially the fact that a comma is ignored after three zeroes to the right :eyes: But this  only arises when users type such large numbers in price or unit.

I think to address this and come up with a more definite solution we could consider improving the consistency of our Currency/Localization throughout the app. Also, I'm not sure changing languages should override the instances settings...

I realize this is not a problem for all instances. Some only have one locale available and the role is perhaps clearer in that case. But in addition to instances which support several locales, there are even countries which support both arithmetic conventions:

image

Do we have merge consensus on this PR?

**Comment by jaycmb** [comment]
@RachL 

>  1,000g = 1g.
Correct.

But 1.000 in European countries is 1000 grams=thousand grams ;)
I was referring to using the` .` which is the common thousand separator in EU

**Comment by jaycmb** [comment]
> > Entering 1.000 (thousand, according to number localisation)
> 
> On your screenshot, this seems to be `1,000` right?

SORRY. I was playing so much back and forth between commas and dots in all fields that I got comma blinded. Updated screenshot.
It handles bot `. `and` , `as a decimal separator, not a thousands separator.
Not sure if we have to worry - time will show ;)

**Pull request merged**